### PR TITLE
feat: activity table filters

### PR DESCRIPTION
**Pull request by joanna-pagepro**
https://tw.auroracreation.com/app/tasks/15669197

- filters and search for activity table on activity page

**Acceptance criteria**
- Filtering
  - Search activity:
    - The search sound query both the action title (in dark bold) and the metadata information including the user who created the action (in bold blue).
    - All of the results shown in the example either have the title, metadata or user within the metadata in the search result.
    - After typing user should be abe to use X button to reset search
  - Filter by:
    - Action type:
      - all availabe action types
      - scroll on desktop/tablet
    - Status:
      - Staked, Objected, Escalated, Voting, Reveal, Finalizable, Passed, Failed
    - Date:
      - past hour, 24 hours, week, month, year
      - Datapicker for custom range
    - Decision method:
      - Multi-sig, Permission, Reputation
  - General requirements:
    - hide pagination if not needed
    - show empty state if theres no results to show

### CODE REVIEW

**Review by melyndav — CHANGES_REQUESTED** [review]
Hey @joanna-pagepro thanks for your work on this one! A couple of feedback points

1. The button title should be 'All filters'

2. The icon inside the button should be 18x18px to match the design system component - https://www.figma.com/file/l1dOM5qiQYwF0ElvKDqqjg/Design-System---Colony-v3?type=design&node-id=1782-13112&mode=design&t=XhbkiQKGmndBtCBQ-4 

3. I created new actions to test but they are not showing when I select the 'past hour' filter
image

Date picker issues below - link to component in design system for reference
https://www.figma.com/file/l1dOM5qiQYwF0ElvKDqqjg/Design-System---Colony-v3?type=design&node-id=1782-15095&mode=design&t=XhbkiQKGmndBtCBQ-4

4. The desktop date picker does not need the x icon showing only for mobile.
image

5. The apply and cancel buttons only show when a range has been selected and default
image

6. Hover states are missing on the years dropdown.

7. Selected dates ssould be round
image

8. When I add a custom date range and open up the date picker again, it doesn't show the active selected range
image

9. There looks to be some padding issues in the date dropdown 

image

10. i created actions today but no results are showing for today's date even when permissions are used and the actions are completed.

image

**Comment by joanna-pagepro** [comment]
@melyndav when no filters are applied - can you see the action that you've created with today's date?

File: src/components/common/ColonyActionsTable/partials/ActionsTableFilters/partials/filters/ActionTypeFilters/ActionTypeFilters.tsx
Comment (jakubcolony) [inline src/components/common/ColonyActionsTable/partials/ActionsTableFilters/partials/filters/ActionTypeFilters/ActionTypeFilters.tsx:10]: Same as the other default value, this might not be needed

**Comment by melyndav** [comment]
> @melyndav when no filters are applied - can you see the action that you've created with today's date?

Yes, the actions show by default.

**Comment by joanna-pagepro** [comment]
> > @melyndav when no filters are applied - can you see the action that you've created with today's date?
> 
> Yes, the actions show by default and no filters selected.

And the date is correct?

File: src/components/v5/common/Fields/Datepicker/Datepicker.tsx
Comment (jakubcolony) [inline src/components/v5/common/Fields/Datepicker/Datepicker.tsx:21]: In `react-datepicker` they use `WithRange` as the name of this generic, could we rename it here too? It'd be easier to figure out what it does

**Comment by melyndav** [comment]
> > > @melyndav when no filters are applied - can you see the action that you've created with today's date?
> > 
> > 
> > Yes, the actions show by default and no filters selected.
> 
> And the date is correct?

Sorry, this is on me. I overlooked the dates and in local dev they are showing as being created in the future (which I can't test in the date picker). So I guess I won't be able to test this until it goes to production. So you can ignore anything related to the specific date. :( 

image

File: src/hooks/useActivityFeed/useActivityFeed.ts
Comment (jakubcolony) [inline src/hooks/useActivityFeed/useActivityFeed.ts:49]: Could you provide a comment about this?

File: src/hooks/useActivityFeed/helpers.ts
Comment (jakubcolony) [inline src/hooks/useActivityFeed/helpers.ts:48]: `colony` should always be defined

**Review by jakubcolony — COMMENTED** [review]
Nice work @joanna-pagepro, I did some functional testing (although I'm missing the specs as there's no linked GH issue) and all seems to be working well.

Code looks good too, I like the separation of concerns and improved clarity over the previous PR. I only left some minor comments.

One question I have is whether the search filter should also allow searching within the action metadata/description? Otherwise searching for `Pay lucy` in this table will yield no results:
image

File: src/components/common/ColonyActionsTable/hooks/useRenderRowLink.tsx
Comment (joanna-pagepro) [reply src/components/common/ColonyActionsTable/hooks/useRenderRowLink.tsx:11]: It's passed to`renderCellWrapper` which type is a function. This field is used in other tables as well and I'm not sure if it's ok to update it in this PR.

File: src/hooks/useActivityFeed/useActivityFeed.ts
Comment (joanna-pagepro) [reply src/hooks/useActivityFeed/useActivityFeed.ts]: I tested something and forgot to undo the changes here.

**Comment by joanna-pagepro** [comment]
@melyndav @jakubcolony The Datepicker component is now in a separate PR (https://github.com/JoinColony/colonyCDapp/pull/1929) and I would like to merge it first, before this PR. so that it won't block other tasks where this component is needed, so all of the fixes related to the form will be added there.

**Comment by joanna-pagepro** [comment]
> Nice work @joanna-pagepro, I did some functional testing (although I'm missing the specs as there's no linked GH issue) and all seems to be working well.
> 
> Code looks good too, I like the separation of concerns and improved clarity over the previous PR. I only left some minor comments.
> 
> One question I have is whether the search filter should also allow searching within the action metadata/description? Otherwise searching for `Pay lucy` in this table will yield no results: image

`The search should query both the action title (in dark bold) and the metadata information including the user who created the action (in bold blue).` - right now we have it implemented to search by the title, but `metadata?.customTitle` returns null and as a fallback the text that is used as a description is used for the title as well. Unfortunately, there is a problem with using that for search, because it's not a `string`, but `React.Node`.

**Comment by joanna-pagepro** [comment]
@melyndav @jakubcolony datepicker merged - please check if filters are ok now
@jakubcolony please also take a look at my previous comment.

File: src/components/common/ColonyActionsTable/hooks/useRenderRowLink.tsx
Comment (jakubcolony) [reply src/components/common/ColonyActionsTable/hooks/useRenderRowLink.tsx:11]: I couldn't see `useRenderRowLink` being used anywhere else and was hoping it could be turned into a component instead of a hook, especially considering @chmanie's proposal in: https://discord.com/channels/562263648173555742/998629903740182580/1207325289227755580
> It's passed to renderCellWrapper which type is a function.

It should still be ok to pass a component to it as it is a function, right?

File: src/components/v5/common/Fields/Datepicker/Datepicker.tsx
Comment (jakubcolony) [inline src/components/v5/common/Fields/Datepicker/Datepicker.tsx:1]: Is this file still needed after #1929 merge?

**Review by jakubcolony — COMMENTED** [review]
Thanks for making changes @joanna-pagepro, code looks good, I'm just unsure if we still need the Datepicker file so it'd be great if you could shine some light on it.

> Unfortunately, there is a problem with using that for search, because it's not a string, but React.Node.

Would it be feasible to just run the metadata through some react node to text library, such as this one - https://www.npmjs.com/package/react-to-text? (I haven't researched it, it's the first match from google)

If not then this needs a separate issue.

File: src/components/v5/common/Fields/Datepicker/Datepicker.tsx
Comment (joanna-pagepro) [reply src/components/v5/common/Fields/Datepicker/Datepicker.tsx]: It's not, you're right. I forgot to remove it after the merge.

**Review by ArmandoGraterol — DISMISSED** [review]
Looking good to me 👌 Nice work!

**Review by melyndav — DISMISSED** [review]
Hey @joanna-pagepro! Thanks for your updates. A few more required.

1. On mobile all filter should be expanded by default. This lines up with the behavior in other filter PR's. 
image
 
 2. Chevron arrows should be text-gray-900 as per Figma in the screenshot below:
 
image


3. The heading on mobile should be 'filter & sort' as per the design in the screenshot below:

image

4. Chevron in the mobile search should be a different size as per the component in Figma:

image
 
 Link to component: https://www.figma.com/file/l1dOM5qiQYwF0ElvKDqqjg/Design-System---Colony-v3?type=design&node-id=1782-13597&mode=design&t=NcS05TF8IrULk8wQ-4
 
 5. Can the mobile empty state for where there are no search results match the correct height in Figma? Currently it's the same height on desktop as it is on mobile and you can't see the no result text clearly. 
 
 As per Figma: 
image

6. There are some alignment issues with the number in the filter box on mobile. 
image

File: src/components/common/ColonyActionsTable/FiltersContext/FiltersContext.tsx
Comment (jakubcolony) [inline src/components/common/ColonyActionsTable/FiltersContext/FiltersContext.tsx:204]: Just coming back to this from #1927, there should be an enum/union type identifying those filters instead of relying on i18n strings

Also the name `category` is a bit unclear, it's the field you're filtering on, right?

**Review by jakubcolony — CHANGES_REQUESTED** [review]
Still looking nice, just duplicating my comment from #1927 about i18n strings as keys.

**Review by ArmandoGraterol — DISMISSED** [review]
LGTM 👍

**Review by iamsamgibbs — CHANGES_REQUESTED** [review]
Thanks for your work on this, it is looking really close!

The page is shifting for me when selecting a filter.

https://github.com/JoinColony/colonyCDapp/assets/34915414/a2ed4a09-06fa-4876-bb48-08053ef27a33

I'm also getting this error in the console.

Screenshot 2024-02-27 at 10 55 17

**Comment by joanna-pagepro** [comment]
@iamsamgibbs page shifting is visible here, because the list of actions is shorter after selecting a filter and because of that there's no scrollbar on the right. This issue will be the same for other pages where there's filtering and I think that it shouldn't be fixed in this PR because of that. There should be a separate issue to take care of it if needed.

**Review by iamsamgibbs — APPROVED** [review]
Ah yep, I see, I agree if we want to tackle that it should be a separate issue. All looks good now!

**Comment by AdrianDudko** [comment]
> Ah yep, I see, I agree if we want to tackle that it should be a separate issue. All looks good now!

Here's an issue for that https://github.com/JoinColony/colonyCDapp/issues/1990 . It's missing a description yet, but has comments linked, so we won't forget what should be done.

**Review by rdig — DISMISSED** [review]
Approving so we can get this merged

**Review by jakubcolony — APPROVED** [review]
Looks good, thanks for addressing all the feedback @joanna-pagepro

**Review by CzarekDryl — APPROVED** [review]

**Review by iamsamgibbs — APPROVED** [review]

**Pull request merged**
